joomla/joomla-cms
# [RFC] [WiP] Add MIME type check for upload files to com_joomlaupdate

Closes #29764 .

Solves also server side part of issue #29763 . Client side part of that issue has been solved with PR #29877 .

### Summary of Changes

This Pull Request (PR) adds a MIME type check for files uploaded with the "Upload & Update" function of the Joomla Update Component.

Currently the "Upload & Update" function supports only zip files, and so only MIME type `application/zip` will be allowed here. No idea what the other update packages (tar.gz, tar.bz2) are good for. There is no update from folder option or update channel for which they could be used.

In opposite to the MIME type checks in the media manager, this check here executes the 2nd check using the `finfo_...` function not only when the `mime_content_type` function is not available, it also executes the 2nd check when the 1st check failed with an exception. This makes more safe against false alarms e.g. due to problems on that server with the magic file used by `mime_content_type`.

### Requests for Comments

**RFC 1:** Should I add a configuration option to the Joomla Update Component for switching off this new MIME type check on hosts where there is some problem with MIME type detection, and extend the language strings shown in case of error or wrong MIME type by corresponding hint that this might be a false alarm on some hosts and can be switched off in the options?

**RFC 2:** Should the same check also added for files downloaded by the "Live Update" function of the Joomla Update Component?

**RFC 3:** Shall such check(s) also be added to the Extensions Installer?

**RFC 4:** Currently many exceptions raised in the function modified by this PRs use language strings from com_installer, except of the new MIME type check added by this PR, which already uses com_joomlaupdate strings. The com_installer strings are currently not translated. This has to be fixed with another PR. How shall it be fixed: Load the com_installer language file, or copy the strings to the com_joomlaupdate language file and rename them from `COM_INSTALLER_MSG_...` to `COM_JOOMLAUPDATE_MSG_...`? And is it a know issue? I didn't find one, but maybe I've missed it.
=> For fixing the untranslated language strings see PR #29899 .

### Testing Instructions

1. On a clean current staging or recent 3.9 nightly build or a 3.9.19, login to backend and go to "Components -> Joomla Update".

2. Go to tab "Upload & Update" and select a file for upload which is not a zip file, e.g. one of the "tar.gz" or "tar.bz2" update packages.
![snap-1](https://user-images.githubusercontent.com/7413183/86286826-78fe6980-bbe7-11ea-8779-4683a375f5fb.png)

3. Use the "Upload & Install" button to start the upload.
Result: You see the login screen for the next update step:
![snap-2](https://user-images.githubusercontent.com/7413183/86286926-a1866380-bbe7-11ea-88a6-7d6f3ba00820.png)

4. Enter the super admin's credentials and use the "Install" button to continue.
Result: You get an error alert at the beginning of the upload:
![snap-3](https://user-images.githubusercontent.com/7413183/86287109-e90cef80-bbe7-11ea-8199-d874d9106dcc.png)

5. Use the "OK" button to close the alert.
Result: You are stuck at the upload start without any progress, see issue #29764 .

6. Use the "Back" button of your broswer to get rid of that page.

7. Apply the patch of this PR.

8. Apply the patch of PR #29881 . That PR is not required for this one here but makes testing easier.

9. Repeat steps 2 and 3.
Result: See section "Expected result AFTER applying this Pull Request" below.

### Actual result BEFORE applying this Pull Request

See testing instructions above. At the end you stick with:
![snap-4](https://user-images.githubusercontent.com/7413183/86287244-22ddf600-bbe8-11ea-8b94-7b5d835e2354.png)

See also issue #29764 .

### Expected result AFTER applying this Pull Request

If both PRs this one here and the PR #29881 have been applied:
![snap-7](https://user-images.githubusercontent.com/7413183/86297438-9fc69b00-bbfb-11ea-9a15-15466219b94b.png)

If only this PR has been applied:
![snap-5](https://user-images.githubusercontent.com/7413183/86287682-e19a1600-bbe8-11ea-91c7-02febaa896af.png)

![snap-6](https://user-images.githubusercontent.com/7413183/86297447-a6eda900-bbfb-11ea-9e02-41345ae79ee6.png)

### Documentation Changes Required

None as far as I know.
